src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+            }
+        }
+
+        impl<'a> fmt::Display for Foo<'a> {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                writeln!(
+                    f,
+                    "{} (nstacks: {}, lines: {})",
+                    self.path.display(),
+                    self.nstacks,
+                    self.nlines
+                )?;
+                let default_duration = self.results[&self.default];
+                let mut results = self.results.iter().collect::<Vec<_>>();
+                results.sort_by(|(_, d1), (_, d2)| (**d1).cmp(*d2));
+                for (nstacks_per_job, duration) in results.iter().take(10) {
+                    writeln!(
+                        f,
+                        "    nstacks_per_job: {:>4} (% of total: {:>3.0}%) | time: {:.0}% of default",
+                        nstacks_per_job,
+                        (**nstacks_per_job as f32 / self.nstacks as f32) * 100.0,
+                        **duration as f64 / default_duration as f64 * 100.0,
+                    )?;
+                }
+                writeln!(f)?;
+                Ok(())
+            }
+        }
+
+        fn count_lines_and_stacks(bytes: &[u8]) -> (usize, usize) {
+            let mut reader = io::BufReader::new(bytes);
+            let mut line = String::new();
+
+            let (mut nlines, mut nstacks) = (0, 0);
+            loop {
+                line.clear();
+                let n = reader.read_line(&mut line).unwrap();
+                if n == 0 {
+                    nstacks += 1;
+                    break;
+                }
+                nlines += 1;
+                if line.trim().is_empty() {
+                    nstacks += 1;
+                }
+            }
+            (nlines, nstacks)
+        }
+
+        let inputs = read_inputs(inputs)?;
+
+        // Warmup
+        let now = Instant::now();
+        stdout.write_fmt(format_args!(
+            "# Warming up for approximately {} seconds.\n",
+            WARMUP_SECS
+        ))?;
+        stdout.flush()?;
+        while now.elapsed() < std::time::Duration::from_secs(WARMUP_SECS as u64) {
+            for (_, bytes) in inputs.iter() {
+                folder.collapse(&bytes[..], io::sink())?;
+            }
+        }
+
+        // Time
+        let mut foos = Vec::new();
+        for (path, bytes) in &inputs {
+            stdout.write_fmt(format_args!("# {} ", path.display()))?;
+            stdout.flush()?;
+            if let Some(foo) = Foo::new(folder, path, bytes, &mut stdout)? {
+                foos.push(foo);
+            }
+            stdout.write(&[b'\n'])?;
+            stdout.flush()?;
+        }
+        stdout.write(&[b'\n'])?;
+        stdout.flush()?;
+        foos.sort_by(|a, b| b.nstacks.cmp(&a.nstacks));
+        for foo in foos {
+            stdout.write_fmt(format_args!("{}", foo))?;
+            stdout.flush()?;
+        }
+
+        Ok(())
+    }
+
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Same here -- I assume these are unchanged?

Reply:
Yes - This was just copy/paste to move all the common functionality into `common.rs`.